Fix  `invoke` return, add `MaxSize` parameter

pmed opened the pull request:
Hi Erik!

I've made a similar solution for `std::function` with fixed size storage some time ago, and I used the same way to implement it. Nice trick with swap in move ctor by the way, I've never seen it before.

I'd like to share a couple of commits. The first commit fixes an error for Functions with non-void return type. And another commit adds an optional `MaxSize` template parameter which controls the final size of Function.

pmed committed:
Fix return value

pmed committed:
Add MaxSize template parameter to control size of Function

MaxSize is used to calculate data storage size with according
to invoker and manager sizes.

rigtorp commented:
Thanks Pavel! 

There is also an issue with storing function pointers to free functions and member functions. I haven't found a solution yet.

pmed commented:
Hi Erik,

You are welcome :)

I actually don't see any issue with member functions. As I understand they should be wrapped to a function object with `std::bind` or with a lambda, and they already work.

I added another pull request #2 that fixes assignment of a reference to free function, and I added more use cases in `example.cpp`
